google-ads-examples/src/main/java/com/google/ads/googleads/examples/planning/ForecastReach.java

```diff
@@ -0,0 +1,352 @@
+        System.out.printf("\t- %s%n", ageRange);
+      }
+      System.out.println("Genders:");
+      for (GenderInfo gender : product.getPlannableTargeting().getGendersList()) {
+        System.out.printf("\t- %s%n", gender.getType());
+      }
+      System.out.println("Devices:");
+      for (DeviceInfo device : product.getPlannableTargeting().getDevicesList()) {
+        System.out.printf("\t- %s%n", device.getType());
+      }
+    }
+  }
+
+  /**
+   * Creates a base request to generate a reach forecast.
+   *
+   * @param customerId the customer ID for the reach forecast.
+   * @param productMix the product mix for the reach forecast.
+   * @param locationId location ID to plan for. To find a valid location ID, either see
+   *     https://developers.google.com/google-ads/api/reference/data/geotargets or call
+   *     ReachPlanServiceClient.ListPlannableLocations.
+   * @param currencyCode three-character ISO 4217 currency code.
+   * @return populated GenerateReachForecastRequest object.
+   */
+  private GenerateReachForecastRequest buildReachRequest(
+      Long customerId, List<PlannedProduct> productMix, String locationId, String currencyCode) {
+    CampaignDuration duration = CampaignDuration.newBuilder().setDurationInDays(28).build();
+
+    List<GenderInfo> genders =
+        Arrays.asList(
+            GenderInfo.newBuilder().setType(GenderType.FEMALE).build(),
+            GenderInfo.newBuilder().setType(GenderType.MALE).build());
+
+    List<DeviceInfo> devices =
+        Arrays.asList(
+            DeviceInfo.newBuilder().setType(Device.DESKTOP).build(),
+            DeviceInfo.newBuilder().setType(Device.MOBILE).build(),
+            DeviceInfo.newBuilder().setType(Device.TABLET).build());
+
+    Targeting targeting =
+        Targeting.newBuilder()
+            .setPlannableLocationId(locationId)
+            .setAgeRange(ReachPlanAgeRange.AGE_RANGE_18_65_UP)
+            .addAllGenders(genders)
+            .addAllDevices(devices)
+            .build();
+
+    // See the docs for defaults and valid ranges:
+    // https://developers.google.com/google-ads/api/reference/rpc/latest/GenerateReachForecastRequest
+    return GenerateReachForecastRequest.newBuilder()
+        .setCustomerId(Long.toString(customerId))
+        .setCurrencyCode(currencyCode)
+        .setCampaignDuration(duration)
+        .setTargeting(targeting)
+        .setMinEffectiveFrequency(1)
+        .addAllPlannedProducts(productMix)
+        .build();
+  }
+
+  /**
+   * Pulls and prints the reach curve for the given request.
+   *
+   * @param reachPlanServiceClient instance of Reach Plan Service client.
+   * @param request an already-populated reach curve request.
+   */
+  private void getReachCurve(
+      ReachPlanServiceClient reachPlanServiceClient, GenerateReachForecastRequest request) {
+    GenerateReachForecastResponse response = reachPlanServiceClient.generateReachForecast(request);
```

Comment:
Can you confirm that this has been tested with the API? I believe that this service requires an allowlist.

Reply:
Confirmed on my side (with the allowlist)

Reply:
Confirmed.